### DISCUSSION TITLE
VolumeSnapshots exceeding available Capacity

### DISCUSSION THREAD

**Discussion by druesendieb** [topolvm/topolvm#958] [Q&A] [answered]
Hi TopoLVM folks,

we're currently facing issues with Snapshots and Storage Capacity Tracking so I hope you can guide us in the correct direction. 

## Situation

We use k8s 1.28.12, Topolvm 0.28 and external-snapshotter v8.0.1.
Our nodes use thin provisioned 14TB volumegroups. 
On each node we have 2 pods using  6,8TB PVCs  for data and 20Gi PVC for config - so we're using almost all disk space. 

This is our deviceClass config:
```
  deviceClasses:
    - name: vg1-thin
      volume-group: vg1
      default: true
      spare-gb: 10
      type: thin
      thin-pool:
        name: vg1-thinpool
        overprovision-ratio: 1.0
```
We deliberately decided to not use overprovisioning, as our volumes will be fully utilized and the application manages data retention on its own. 

These PVCs should be backupped with Kasten K10 (which completely relies on k8s VolumeSnapshot & Restore mechanic). After taking a VolumeSnapshot, Kasten creates a new PVC using the SnapshotContent, which afterwards should be used by a pod to be moved to the backup repository. 

This was tested successfully with small volumes, but with the big volumes our problems start.

## Problems

In this scenario we only backup 1 of the 2 pods/big PVCs as a first test.
The VolumeSnapshot of the first pods 2 PVCS was created successfully by Kasten. 

```
# lvs
  LV                                   VG        Attr       LSize    Pool         Origin                               Data%  Meta%  Move Log Cpy%Sync Convert
  ubuntu-lv                            ubuntu-vg -wi-ao---- <221.99g                                                                                          
  07b50196-b458-487c-a5f3-a7889d484563 vg1       Vri-a-tz--    6.64t vg1-thinpool 4d81a676-fada-4682-b387-9d0cdb312533 81.04                                  
  163dce5a-45c5-4aa7-9f1f-a3cfb82ea090 vg1       Vwi-aotz--    6.64t vg1-thinpool                                      77.05                                  
  4d81a676-fada-4682-b387-9d0cdb312533 vg1       Vwi-aotz--    6.64t vg1-thinpool                                      81.04                                  
  80115051-37a0-4d4d-a5aa-50fefc17ca3f vg1       Vwi-aotz--   20.00g vg1-thinpool                                      20.94                                  
  b0cd8aac-555c-41d6-a5fb-de43605686c8 vg1       Vwi-aotz--   20.00g vg1-thinpool                                      21.41                                  
  ef14c104-a870-4ee8-88d3-6704bc0e5848 vg1       Vwi-a-tz--   20.00g vg1-thinpool                                      21.41                                  
  f72c8edd-2927-4a73-8e08-c2653b2b2f74 vg1       Vri-a-tz--   20.00g vg1-thinpool b0cd8aac-555c-41d6-a5fb-de43605686c8 21.41                                  
  longhorn-lv                          vg1       -wi-ao----  750.00g                                                                                          
  vg1-thinpool                         vg1       twi-aotz--   13.82t                                                   76.02  39.97  
```

As the nodes are already nearly full, this capacity also gets populated to the scheduler. 

```
❯ kg csistoragecapacities.storage.k8s.io --no-headers | awk {'print $1'} | xargs kubectl describe csistoragecapacities.storage.k8s.io | grep Capacity:
Capacity:     513Gi
Capacity:     513Gi
Capacity:     493Gi
Capacity:     513Gi
Capacity:     513Gi
Capacity:     513Gi

❯ kd no node2 | grep capacity
Annotations:        capacity.topolvm.io/00default: 18446737258670194688
                    capacity.topolvm.io/vg1: 550829555712
                    capacity.topolvm.io/vg1-thin: 18446737258670194688
```

I found the capacity numbers of the node annotation weird, so i looked into the topolvm-controller logs
```
W0902 10:33:19.431858 1 capacity.go:552] Retrying capacity.workItem{segment:(*topology.Segment)(0xc000e80df8), storageClassName:"topolvm"} after 14 failures
E0902 10:32:19.431222 1 capacity.go:551] CSI GetCapacity for {segment:0xc000e80df8 storageClassName:topolvm}: rpc error: code = Internal desc = strconv.ParseInt: parsing "18446737258670194688": value out of range
```

Meanwhile Kasten creates a Pod for backup to the repository. This happens once for each PVC/VolumeSnapshot. The pod for the 20Gi PVC was created successfully.
The pod for the 6800Gi PVC can't be scheduled thanks to storage capacity tracking (493Gi < 6800Gi)
```
Unable to schedule pod; no fit; waiting" pod="kasten/copy-vol-data-8jclb" err="0/6 nodes are available: 6 node(s) did not have enough free storage. preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling.."
```

## Questions

- Is this wanted behaviour with storage capacity tracking?
- Would this change if over provisioning is activated? My guess would be that this is only considered for LV creation on the node 
- How can we achieve the goal of backupping these volumes?
- The `value out of range` error seems unintentional to me, might this be a bug?

**Comment by ushitora-anqou** [comment]
Thank you for your feedback. I was able to reproduce your issue. After some investigation, I found that this is caused by a combination of the following reasons:

1. The `overprovision-ratio` takes all the thin volumes, including snapshots, into account. So, when it's set to 1.0, the total size of all the thin volumes must be smaller than the pool size, which I assume is 14TB in your case.
2. TopoLVM has a bug that annotates a node with the wrong capacity around $2^{64}$ when the total size of the thin volumes exceeds the overprovisioned pool size, i.e., `overprovision-ratio` times the VG size.

I'll submit a PR to fix the bug mentioned in 2. Once it's fixed, you won't be able to create a new snapshot if the total size of the thin volumes exceeds the overprovisioned pool size.

To resolve the issue in 1, you need to set the `overprovision-ratio` to greater than 1. In your case, I think 2.0 should be sufficient because, if I understand your situation correctly, each thin volume should have at most one snapshot, so the total size of the thin volumes shouldn't be more than twice the pool size.

And here are the answers to your questions:

> Is this wanted behaviour with storage capacity tracking?

I think Storage Capacity Tracking doesn't work properly in your environment because the capacities annotated in Node objects are wrong. Once `overprovision-ratio` is adjuested and TopoLVM annotates them correctly, it should work normally.

> Would this change if over provisioning is activated? My guess would be that this is only considered for LV creation on the node

`overprovision-ratio` is used not only when a LV is created, but also when a Node object is annotated with its capacity.

> How can we achieve the goal of backupping these volumes?

As I mentioned earlier, you need to make `overprovision-ratio` greater than 1.

> The value out of range error seems unintentional to me, might this be a bug?

Yes, it's definitely a bug. I'll fix it.

**Reply by druesendieb** [reply]
Thank you for your response @ushitora-anqou! 
Looking forward to the PR and the fix. 
We will have a look the bigger overprovision-ratio and a way to monitor physical capacity not being exceeded.

As a side note I have to say, i really enjoy working with TopoLVM - its a solid solution and you are always very helpful with questions and bug fixes!

**Reply by ushitora-anqou** [reply, marked as answer]
I've fixed the bugs in #961 and #962. These fixes will be included in the next TopoLVM's release, which is scheduled for October. I hope you find them useful!

**Reply by druesendieb** [reply]
Thanks a lot @ushitora-anqou, indeed we do!

**Comment by BirknerAlex** [comment]
Is there are way to adjust the `overprovision` value once the LVM Cluster is created and has existing volumes? If yes, how can I change the value without losing existing data. Updating the existing LVMCluster object seems not having an effect.

**Reply by druesendieb** [reply]
As far as I understood, you simply adopt the overprovision-ratio ration for lvmd and update the chart. To quote @ushitora-anqou:
> `overprovision-ratio` is used not only when a LV is created, but also when a Node object is annotated with its capacity.

**Reply by ushitora-anqou** [reply]
@BirknerAlex Hi. I think you're referring to the LVMCluster custom resource from the lvm-operator. Unfortunately, I can't provide any useful information about the lvm-operator since it's not a TopoLVM project. Maybe you could reach out to the maintainers of the lvm-operator for help?
